src/vos/vos_obj.c

```diff
@@ -1928,26 +1938,44 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
 	krec = rbund.rb_krec;
 	umm = vos_obj2umm(oiter->it_obj);
 
-	if (!vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
-				 &oiter->it_punched, &oiter->it_ilog_info))
+	punched = vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
+				      &oiter->it_punched, &oiter->it_ilog_info);
+
+	bmap = krec->kr_bmap;
+	if (oiter->it_iter.it_for_purge && bmap & KREC_BF_AGG_OPT) {
+		if ((bmap & KREC_BF_AGG_NEEDED) == 0) {
+			if (!full_scan && !punched && oiter->it_punched.pr_epc == 0)
```

Comment:
It would be great to add a comment to explain the 'pr_epc == 0' here, and I still don't quite see: if the key is punched at a higher level, won't the whole tree already being moved to GC?

Reply:
Oh, nm. I see that the tree won't be moved to GC if it's still referenced by snapshots. Some comments could be helpful, thanks!

Reply:
Added a comment